README.md
# i-MSCP SGW_Blocker Plugin

Plugin that uses fail2ban to block bad actors and, when blocked, displays a "You are blocked" page.

See [Changelog](CHANGELOG.md) for a detailed description of what has changed in each version.

>Disclaimer: This plugin is not likely to be that well supported.  It works for me, for what I need, right now. YMMV.

Testing and bug reports are welcome.

## Requirements

* i-MSCP 1.5.3-maintenance (plugin version v0.x)

Note that the plugin managemnet in i-MSCP 1.5.3-maintenance may be broken awaiting a merge of the following pull-requests to fix the issues:
* [Fix PluginManager](https://github.com/i-MSCP/imscp/pull/108)
* [Add _ to plugin name validation](https://github.com/i-MSCP/imscp/pull/109)

The plugin has been tested with the i-MSCP 1.5.3-maintenance-sgw branch [available here](https://github.com/saygoweb/imscp).

## Installation

Download the SGW_Blocker.tgz release file from the releases page.  Don't download the Git source archive if
you plan to upload using the plugin management interface. 

1. Upload the plugin through the plugin management interface
2. Install the plugin through the plugin management interface

## How to Help

* Report issues you find in our [GitHub Issue Tracker](https://github.com/saygoweb/imscp-letsencrypt/issues). Please report with as much detail as you can. Simply saying "It doesn't work" will gain you sympathy, but not a lot else.
* Want to contribute code? Go right ahead, fork the project on GitHub, pull requests are welcome.

## License

```
i-MSCP  SGW_Blocker plugin
Copyright (C) 2021 Cambell Prince <cambell.prince@gmail.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; version 2 of the License

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
```

See [LICENSE](LICENSE)

## Authors

* Cambell Prince <cambell.prince@gmail.com>


CHANGELOG.md
# iMSCP SGW_Blocker Plugin - Changelog

## Version 0.1.0

* Proof of concept


test/nat_ban.sh
#!/bin/bash
#iptables -t nat -I f2b-test 1 -s 192.168.33.1 -j REDIRECT --to-port 8890
iptables -t nat -I f2b-test 1 -p tcp -s $1 -j REDIRECT --to-port 8890

test/nat_stop.sh
#!/bin/bash
iptables -t nat -F f2b-test
iptables -t nat -X f2b-test
iptables -t nat -F PREROUTING


test/nat_start.sh
#!/bin/bash
# iptables -t nat -F PREROUTING
iptables -t nat -N f2b-test
# iptables -t nat -A f2b-test -p tcp -j REDIRECT --to-port 8890
iptables -t nat -I PREROUTING -p tcp --dport 80 -j f2b-test

config.php
<?php
/**
 * i-MSCP Blocker plugin
 * Copyright (C) 2021 Cambell Prince <cambell.prince@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

return array(
);


SGW_Blocker.php
<?php
namespace iMSCP\Plugin\SGW_Blocker;
/**
 * i-MSCP Blocker plugin
 * Copyright (C) 2021 Cambell Prince <cambell.prince@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

use iMSCP\Event\Event;
use iMSCP\Event\EventManagerInterface;
use iMSCP\Event\Events;
use iMSCP\Plugin\AbstractPlugin;
use iMSCP\Plugin\PluginException;
use iMSCP\Plugin\PluginManager;
use iMSCP\Registry;

/**
 * Class SGW_Blocker
 */
class SGW_Blocker extends AbstractPlugin
{
    /**
     * Plugin initialization
     *
     * @return void
     */
    public function init()
    {
        l10n_addTranslations(__DIR__ . '/l10n', 'Array', $this->getName());
    }

    /**
     * Register event listeners
     *
     * @param EventManagerInterface $eventsManager
     * @return void
     */
    public function register(EventManagerInterface $eventsManager)
    {
        // TODO Also sub domains? CP 2017-06
        $eventsManager->registerListener(
            array(
                Events::onResellerScriptStart, // Don't think we need this except to enable it at all maybe CP 2017-06
                Events::onClientScriptStart,
                Events::onAfterDeleteDomainAlias, // Cleanup Blocker for the domain alias
                Events::onAfterDeleteCustomer     // Cleanup Blocker for this domain / customer
            ),
            $this
        );
    }

    /**
     * Plugin installation
     *
     * @throws PluginException
     * @param PluginManager $pluginManager
     * @return void
     */
    public function install(PluginManager $pluginManager)
    {
        try {
            $this->migrateDb('up');
        } catch (PluginException $e) {
            throw new PluginException($e->getMessage(), $e->getCode(), $e);
        }
    }

    /**
     * Plugin update
     *
     * @throws PluginException When update fail
     * @param PluginManager $pluginManager
     * @param string $fromVersion Version from which plugin update is initiated
     * @param string $toVersion Version to which plugin is updated
     * @return void
     */
    public function update(PluginManager $pluginManager, $fromVersion, $toVersion)
    {
        try {
            $this->migrateDb('up');
            $this->clearTranslations();
        } catch (PluginException $e) {
            throw new PluginException($e->getMessage(), $e->getCode(), $e);
        }
    }

    /**
     * Plugin uninstallation
     *
     * @throws PluginException
     * @param PluginManager $pluginManager
     * @return void
     */
    public function uninstall(PluginManager $pluginManager)
    {
        try {
            $this->migrateDb('down');
            $this->clearTranslations();
        } catch (PluginException $e) {
            throw new PluginException($e->getMessage(), $e->getCode(), $e);
        }
    }

    /**
     * onResellerScriptStart event listener
     *
     * @return void
     */
    public function onResellerScriptStart()
    {
        $this->setupNavigation('reseller');
    }

    /**
     * onClientScriptStart event listener
     *
     * @return void
     */
    public function onClientScriptStart()
    {
        // if (self::customerHasBlocker($_SESSION['user_id'])) { // TODO CP 2017-06
            $this->setupNavigation('client');
        // }
    }

    /**
     * onAfterDeleteCustomer event listener
     *
     * @param Event $event
     * @return void
     */
    public function onAfterDeleteCustomer(Event $event)
    {
    }

    /**
     * onAfterDeleteDomainAlias event listener
     *
     * @param Event $event
     * @return void
     */
    public function onAfterDeleteDomainAlias(Event $event)
    {
    }

    /**
     * Get routes
     *
     * @return array
     */
    public function getRoutes()
    {
        $pluginDir = $this->getPluginManager()->pluginGetRootDir() . '/' . $this->getName();
        return [];
    }

    /**
     * Get status of item with errors
     *
     * @return array
     */
    public function getItemWithErrorStatus()
    {
        $stmt = exec_query(
            "
                SELECT letsencrypt_id AS item_id, cert_name AS item_name,
                    'letsencrypt' AS `table`, 'status' AS field
                FROM letsencrypt WHERE status NOT IN(?, ?, ?, ?, ?, ?, ?)
            ",
            array('ok', 'disabled', 'toadd', 'tochange', 'toenable', 'todisable', 'todelete')
        );

        if ($stmt->rowCount()) {
            return $stmt->fetchAll(\PDO::FETCH_ASSOC);
        }

        return array();
    }

    /**
     * Set status of the given plugin item to 'tochange'
     *
     * @param string $table Table name
     * @param string $field Status field name
     * @param int $itemId Blocker item unique identifier
     * @return void
     */
    public function changeItemStatus($table, $field, $itemId)
    {
        if ($table == 'letsencrypt' && $field == 'status') {
            exec_query('UPDATE letsencrypt SET `status` = ? WHERE letsencrypt_id = ?', array('tochange', $itemId));
        }
    }

    /**
     * Return count of request in progress
     *
     * @return int
     */
    public function getCountRequests()
    {
        $stmt = exec_query(
            'SELECT COUNT(letsencrypt_id) AS cnt FROM letsencrypt WHERE `status` IN (?, ?, ?, ?, ?)',
            array('toadd', 'tochange', 'toenable', 'todisable', 'todelete')
        );

        $row = $stmt->fetchRow(\PDO::FETCH_ASSOC);

        return $row['cnt'];
    }

    /**
     * Does the given customer has Blocker feature activated?
     *
     * @param int $customerId Customer unique identifier
     * @return bool
     */
    public static function customerHasBlocker($customerId)
    {
        static $hasAccess = NULL;

        return true; // TODO Currently everyone has Blocker available on their account.

        // if (NULL === $hasAccess) {
        //     $stmt = exec_query(
        //         '
        //             SELECT COUNT(admin_id) as cnt FROM letsencrypt INNER JOIN admin USING(admin_id)
        //             WHERE admin_id = ? AND admin_status = ?
        //         ',
        //         array($customerId, 'ok')
        //     );

        //     $row = $stmt->fetchRow(PDO::FETCH_ASSOC);
        //     $hasAccess = (bool)$row['cnt'];
        // }

        // return $hasAccess;
    }

    /**
     * Inject Blocker links into the navigation object
     *
     * @param string $level UI level
     */
    protected function setupNavigation($level)
    {
        if (Registry::isRegistered('navigation')) {
            /** @var Zend_Navigation $navigation */
            $navigation = Registry::get('navigation');

            if ($level == 'reseller') {
                if (($page = $navigation->findOneBy('uri', '/reseller/users.php'))) {
                    $page->addPage(array(
                        'label'              => tr('Blocker'),
                        'uri'                => '/reseller/letsencrypt.php',
                        'title_class'        => 'users',
                        'privilege_callback' => array(
                            'name' => 'resellerHasCustomers'
                        )
                    ));
                }
            } elseif ($level == 'client') {
                if (($page = $navigation->findOneBy('uri', '/client/domains_manage.php'))) {
                    $page->addPage(array(
                        'label'       => tr('Blocker'),
                        'uri'         => '/client/letsencrypt.php',
                        'title_class' => 'domains',
                        'pages'       => array(
                            'letsencrypt_edit' => array(
                                'label'       => tr('Blocker Edit'),
                                'uri'         => '/client/letsencrypt_edit.php',
                                'title_class' => ''
                            )
                        )
                    ));
                }
            }
        }
    }

    /**
     * Clear translations if any
     *
     * @return void
     */
    protected function clearTranslations()
    {
        /** @var Zend_Translate $translator */
        $translator = Registry::get('translator');

        if ($translator->hasCache()) {
            $translator->clearCache($this->getName());
        }
    }
}


info.php
<?php
/**
 * i-MSCP SGW_Blocker plugin
 * Copyright (C) 2021 Cambell Prince <cambell.prince@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

return array(
    'author' => 'Cambell Prince',
    'email' => 'cambell.prince@gmail.com',
    'version' => '0.1.0',
    'require_api' => '1.5.1',
    'date' => '2021-08-05',
    'name' => 'SGW_Blocker',
    'desc' => 'Plugin that redirects fail2ban blocks to a web page.',
    'url' => 'https://github.com/saygoweb/imscp-plugin-blocker'
);
